### ISSUE TITLE
MDCMenu: Parameterize whether menu automatically focuses the first item

### ISSUE THREAD

**Issue by KNawm** [material-components/material-components-web#3922]
### What MDC Web Version are you using?

0.40.1

### What browser(s) is this bug affecting?

Tested on Chrome, Chrome for Android and Firefox

### What OS are you using?

Windows 10

### What are the steps to reproduce the bug?

1. Run the demo server
2. Go to localhost:8080/menu.html
3. Open any menu
4. Observe the component's behavior

### What is the expected behavior?

In Material examples of behavior it doesn't selects the first item in a menu on opening.
![](https://i.imgur.com/o83ucou.gif)

### What is the actual behavior?

It automatically selects the first item
![](https://i.imgur.com/w1L3352.gif)

### Any other information you believe would be useful?

I don't know if this is intended behavior but I tested opening a menu and navigating with the keyboard arrows and if it is intended it should be blank and only select the first item if a user uses the keys or other form of navigating to select the appropriate option.

**Comment by MichaelTamm** [comment]
Related bug: Although the focus can be moved via arrow keys, once the menu is open, pressing the [Tab] key closes the menu. (Expected behaviour: The next menu item should be focused.)

**Comment by acdvorak** [comment]
@KNawm: Thanks for filing this issue! We'll reach out to Design to get clarification.

@MichaelTamm: The behavior you're seeing—where pressing the <kbd>Tab</kbd> key closes the menu—is actually intentional. Menus are only supposed to have one tab stop, and the <kbd>↑</kbd> and <kbd>↓</kbd> arrow keys are used to focus the previous/next menu item. This mirrors how menus behave in Google Docs, macOS, and Windows, for example.

**Comment by AlexZarivayskiy** [comment]
Perhaps, this is a temporary solution, until that issue will be solved.
```
HTML
...
<!-- Add my-menu-class -->
<div class="mdc-menu mdc-menu-surface my-menu-class" tabindex="-1">
...


CSS
.my-menu-class :not(.mdc-list--non-interactive)>:not(.mdc-list-item--disabled).mdc-list-item::before, 
.my-menu-class :not(.mdc-list--non-interactive)>:not(.mdc-list-item--disabled).mdc-list-item::after
{
	background-color: #fff;
}
.my-menu-class .mdc-list-item:hover {
	background-color: #f0f0f0;
}
```

**Comment by kfranqueiro** [comment]
I talked with design today, and MDC Menu on Web is intended to align with the [ARIA specification](https://www.w3.org/TR/wai-aria-practices/#menu).

The spec is far more detailed and contextual than simply "yes" or "no" regarding immediately focusing a menu item, but focusing a menu item by default when it opens is acceptable to most if not all cases, whereas not focusing a menu item by default would be detrimental especially to cases involving a single menu opened via a standalone button.

To explain the spec images, the mobile example reflects native mobile UIs, where MDC has a tendency to not consider focus state at all. The desktop example is probably a little bit off, but what should be effectively happening in that case is the menu bar item remains focused until you use arrow keys to navigate into the menu.

Ultimately, I think we will need to parameterize the `open` API for Menu (which probably means it can't simply be a getter/setter anymore) to allow indicating whether to immediately focus an item or not, to make it possible for developers to implement more advanced treatment for e.g. menu bars.

**Comment by KNawm** [comment]
First of all thanks for the response, and while it is great that you will expand the `open` API for Menu it doesn't address the problem entirely, my thought after reading the ARIA spec is that the Menu component does behave strangely in comparison with other components and not in the correct way.

First and foremost the ARIA spec doesn't cover behavior using mouse interaction, it takes it for granted.

I'm going to start with some interactions to look how they differentiate from the current Menu behavior
## Real-world Examples
###### Notepad - Windows 10
![](https://i.imgur.com/HISpQKq.gif)
- Using mouse: Does **NOT** focus the first element, if you press arrow down it will focus the first element.
- Using keyboard: Does focus the first element, if you press arrow down it will focus the second element.

###### Menu bar - MacOS 10.14 (Mojave)
![](https://i.imgur.com/4f9lbrg.gif)
- Using mouse: Does **NOT** focus the first element, if you press arrow down it will focus the first element.
- Using keyboard: Does **NOT** focus the first element, if you press arrow down it will focus the first element.

###### Contextual menu - Windows 10
![](https://i.imgur.com/lfiBRYC.gif)
- Using mouse: Does **NOT** focus the first element, if you press arrow down it will focus the first element.
- Using keyboard: Does **NOT** focus the first element, if you press arrow down it will focus the first element.

So we can see that neither Windows or MacOS places focus on the first element when opening a menu with a mouse click, sometimes not even using the keyboard.

But then we can read this in the [ARIA spec](https://www.w3.org/TR/wai-aria-practices/#menu):
> When a menu opens, or when a menubar receives focus, keyboard focus is placed on the first item. All items are focusable as described in 5.6 Keyboard Navigation Inside Components.

But it is under a headline called **"Keyboard Interaction"** so this applies in general or only with keyboard input? It's not clear and leaves it to interpretation.

> Except where noted, menus opened from a menubutton behave the same as menus opened from a menubar.

Since both menus (and other types of menus) should behave the same way and the ARIA spec has examples for both of them then it should be easy to see what is the correct behavior right? Well not quite...

## ARIA Menubar Examples
###### [Navigation Menubar Example](https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html)
![](https://i.imgur.com/kOEvtT3.gif)
- Using mouse: Does **NOT** focus the first element, if you press arrow down it will focus the first element if it was focused previously as it is a hover menu.
- Using keyboard: Does focus the first element, if you press arrow down it will focus the second element.

###### [Editor Menubar Example](https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html)
![](https://i.imgur.com/fhxYswi.gif)
- Using mouse: Does **NOT** focus the first element, if you press arrow down it will focus the first element.
- Using keyboard: Does focus the first element, if you press arrow down it will focus the second element.

## ARIA Menu Button Examples
These are trickier because they open on hover AND mouse click (why? go figure)
It seems like they were planned as hover buttons but with a poor execution.
###### [Navigation Menu Button Example](https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html)
![](https://i.imgur.com/IYnipMb.gif)
- Using mouse:
1. On hover: Does **NOT** focus the first element, if you press arrow down it will focus the first element if it was focused previously.
2. On mouse click: Does focus the first element, if you press arrow down it will focus the first element.
- Using keyboard: Does focus the first element, if you press arrow down it will focus the second element.

###### [Actions Menu Button Example](https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions.html)
![](https://i.imgur.com/vwCkrXt.gif)
- Using mouse:
1. On hover: Does **NOT** focus the first element, if you press arrow down it will focus the first element if it was focused previously.
2. On mouse click: Does focus the first element, if you press arrow down it will focus the first element.
- Using keyboard: Does focus the first element, if you press arrow down it will focus the second element.

These examples were taken from the latest published version but I checked the latest editor's draft and even [ARIA repo](https://github.com/w3c/aria-practices) to find other versions but all behaved the same way.
Doing that I came across [this](https://github.com/w3c/aria-practices/issues/598) issue where one user quoted one of the "Keyboard Interaction" bullet-points that I cited earlier:
> When a menu opens, or when a menubar receives focus, keyboard focus is placed on the first item. All items are focusable as described in 5.6 Keyboard Navigation Inside Components.

So @mcking65 (Editor of the [ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices/)) answered:
> This text is in the keyboard interaction section, describing only keyboard interaction. That is, the focus movement should occur only as a result of the keyboard commands that can open a menu.

While it isn't explicitly says that the first element shouldn't be focused on mouse click it specifically says the content from the menu's "Keyboard Interaction" section covers just that, keyboards. It is logical that if a user was using a mouse as input it is very likely to keep doing just that, or how many times did you opened a menu with the mouse to select an item with the keyboard?

It is pretty clear that ARIA specs were misunderstood (not surprisingly given the lack of clarity) but it is important to note what @mcking65 said in the same [comment](https://github.com/w3c/aria-practices/issues/598#issuecomment-365168929):
> Providing mouse interaction guidance is not containable and is not in the APG roadmap.

I personally think that it is evident that even in ARIA's examples it states that focus should **NOT** be put in the first element with mouse interaction but it is perfectly great to do it with keyboard interaction.
So maybe can Design have another look?

**Comment by kfranqueiro** [comment]
Thanks for the thorough reply. I came to a lot of similar conclusions/observations yesterday while looking into this (especially the surprising lack of detail RE mouse interactions and the baffling hover behavior in their menu button examples). I should've thought to dump more of those details in my previous reply, sorry about that.

The reason I scoped this issue to parameterizing our Menu component is because a lot of this behavior  is tied to interactions with the element that causes the menu to open, not the menu itself. If someone were to theoretically implement a menu bar component (I say theoretically since there isn't currently one on the Material roadmap), it would need to tell the menu whether to immediately focus an item based on exactly what interaction was performed on the menu bar item. The goal of parameterizing the API is to make the problem solvable to begin with.

I don't think we can go to the extent of attaching event listeners to the anchor directly from the menu component (this was something Abhinay suggested yesterday), because some of the interaction behaviors depend on what type of component the menu is being opened from - and can even be dependent on siblings of the anchor (e.g. arrow left or right from a neighboring menu bar item). Trying to bake in conditions for all of this directly into Menu is likely to add a lot of complexity, vs. handling it at the level of the component responsible for opening the menu.

**Comment by KNawm** [comment]
I see, it obviously wouldn't make sense to put all the conditions in the menu, it would make it over complex and too dependent of the other components effectively damaging modularity, thanks for the response and keep up the good work!

**Comment by mcking65** [comment]
First, I'm glad people are reading the ARIA Practices so closely.

Yes, there are shortcomings in mouse behavior in some of the examples. The only example where these issues have been thoroughly addressed is the editor menubar example. If you follow the related issues link at the top of the example page, you will see we have received a lot of feedback and have issues to fix.

That said, it's really, really important to note that menus that open from a button click are quite different from menubars.

When desktop screen reader users press enter or space (ctrl-opt-space on Mac) to activate a menu button when there screen reader is in its default reading mode, the screen reader calls a method for executing the default action for the button using a system-level accessibility API. That method generates a mouse click event at the center of the button. When the menu opens, if DOM focus does not move to the first menuitem, the screen reader will not be aware that the menu opened. Further, if the screen reader user happens to hope that maybe the menu opened and goes looking for the menuitems, the user is not likely to find the menuitems unless the first menuitem is an immediate DOM sibling of the button.

Unfortunately, this legacy behavior of the system-level accessibility APIs is not going to change in the near term. There are new APIs in the works that could provide some relief in the future. But for now, your code must assume that most screen reader users will appear to be mouse users until they are navigating inside the menu, at which time, they will then appear to be keyboard users and need the menu widget to provide arrow key support.

**Comment by CiriousJoker** [comment]
Is there a possibility, that you guys implement a way for us to completely disable focus grabbing? Even when I set the list to non-interactive, the list still grabs the initial focus.

I have created a select box component out of an mdc-textfield and an mdc-menu. Now, when I click on the textfield, the focus goes to the textfield. At the same time, the menu opens with a list inside. This takes the focus away from the textfield. When I want to focus the textfield again, the menu closes. "Solving" this programmatically by giving focus back to the textfield after opening the menu has weird visual bugs and would introduce all kinds of problems down the road...
